### PR TITLE
#12883: Add initial unit tests for N300

### LINKED ISSUE #12883: Add Initial N300 CCL unit tests
We are missing (even basic) test coverage for CCLs on N300. Although our testing on t3000 should naturally have high overlap with n300, it can't ensure the basics don't regress (e.g. basic behaviour to run all all-gathers on n300 as line).

Ideally these tests can give us sanity-test level coverage in post commit and for that reason I think we can add a very small number of tests (2 per op) and leave it there. Over time we may swap these specific cases out for more complex ones but we should try to have atleast 1 or 2 in. 

This will also help catch accidental CCL related regressions in case someone unknowingly breaks something in CCL ops which as of now are only tested in frequent and nightly jobs.

I propose we start with these 2 test cases on n300:

## 1
interleaved
2 chips
1 link
shape (before fracture) [1, 1, 64, 16384]
dim=3
layout = ttnn.TILE_LAYOUT),

## 2
input_shape (per chip): [1, 1, 512, 2048],
shard shape: [128, 256]
Shard strategy: width
Core range:            ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(7, 3))}),
2 chips
1 link
row major orientation

## Calling All Gather
When calling all-gather, we shouldn't distinguish between line or ring and should just call default `ttnn.all_gather`

## Test Location
I would like to put this under ttnn-unit tests as a new ttnn-group (currently the highest group number is `ttnn group 6`. Pending @tt-rkim's approval, we should make a new one: `ttnn ccl group`. This is specified in `.github/workflows/ttnn-post-commit.yaml`

### PR DESCRIPTION
**Pull request by Aswinmcw**
Tracking Issue : #12883 

Currently we don't have any tests for CCL to be run in N300, this PR will add initial tests to be run as post commit

### Checklist
- [x] Post commit CI passes - https://github.com/tenstorrent/tt-metal/actions/runs/10955272562
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/test_all_gather_N300_post_commit.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+import torch
+import pytest
+from loguru import logger
+import ttnn
+from tests.tt_eager.python_api_testing.sweep_tests.comparison_funcs import comp_equal, comp_pcc
+from models.utility_functions import skip_for_grayskull
+
+
+def is_unsupported_case(input_shape, dim, mem_config, num_devices, num_links, input_dtype, layout):
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_all_gather_N300_post_commit.py:13]: This looks like it's still the same check as in test_all_gather.py. Please import and use that one instead.

File: tests/ttnn/unit_tests/operations/test_all_gather_N300_post_commit.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+            f"Input shape {input_shape} incompatible with {num_devices} on dim {dim} because some chips will have no tensor",
+        )
+
+    if input_shape == [8, 8, 256, 384] and dim == 1 and layout == ttnn.TILE_LAYOUT and input_dtype == ttnn.bfloat8_b:
+        return True, "Known failure"
+
+    return False, ""
+
+
+def run_all_gather_on_n300_impl(
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_all_gather_N300_post_commit.py:55]: Can we commonize this with the main run function from test all-gather? Really the only difference should be the device count check - everything else I believe should be commonizable.

File: tests/ttnn/unit_tests/operations/test_all_gather_N300_post_commit.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    enable_async,
+    n_worker=None,
+    n_buffer=None,
+    num_iter=1,
+    trace_mode=False,
+):
+    if len(all_devices) != 2:
+        pytest.skip("Not N300!")
+
+    for device in all_devices:
```
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_all_gather_N300_post_commit.py:198]: Same commonization comment

File: ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp [hunk 4]
```diff
@@ -138,9 +138,7 @@ void AllGather::validate(const std::vector<Tensor> &input_tensors) const {
     TT_FATAL(this->num_links > 0, "Error");
     TT_FATAL(this->num_links <= input_tensor.device()->compute_with_storage_grid_size().y, "Worker cores used by links are parallelizaed over rows");
     TT_FATAL(this->receiver_device_id.has_value() || this->sender_device_id.has_value(), "Error");
-    if (this->receiver_device_id == this->sender_device_id) {
-        TT_FATAL(input_tensor.device()->get_ethernet_sockets(this->receiver_device_id.value()).size() >= 2 * this->num_links, "2 Device all gather requires at least 2 eth connections per link");
-    } else {
+    if (this->receiver_device_id != this->sender_device_id) {
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp:141]: We should just be able to delete the condition. For 2 chip, we should always have topology linear anyways, so the checks inside should always pass.

**Review by SeanNijjar — CHANGES_REQUESTED** [review]

**Comment by SeanNijjar** [comment]
I forgot to mention in my review the but tests have not yet been enabled in the post-commit run in your PR. Please add that.

**Review by SeanNijjar — CHANGES_REQUESTED** [review]

**Review by SeanNijjar — APPROVED** [review]
Thanks,  this looks better now! Please also run t3000 frequent and nightly pipeline since the functions run during those tests have been modified.

File: ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp [hunk 5]
```diff
@@ -193,7 +187,7 @@ Tensor all_gather(
     all_gather_op::Topology topology = all_gather_op::Topology::Ring;
     auto devices = input_tensor.get_workers();
     uint32_t num_devices = devices.size();
-    if (num_devices == 1){
+    if (num_devices == 2){
         topology = all_gather_op::Topology::Linear;
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp:191]: I don't fully understand here why we check for 2 specifically?   Why not > 2 ?

**Review by eyonland — APPROVED** [review]
